---
title: "Getting Started with STcompare"
output: rmarkdown::html_vignette
author: "Vivien Jiang" 
vignette: >
  %\VignetteIndexEntry{getting-started-with_STcompare}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


# Introducton 

Spatial transcriptomic (ST) technologies enable the investigation of how tissue organization relates to cellular function by profiling the spatial location of cells within a tissue and the gene expression profiles associated with those locations. As ST technologies are increasingly applied in the context of disease characterization and translational research, identifying genes that spatially change in their expression patterns in diseased tissues as compared to healthy controls could reveal localized molecular variation associated with pathological processes, offering insights into disease mechanisms as well as aiding in the discovery of diagnostic biomarkers. 

Attempts at such comparative analysis of ST datasets can be performed using traditional non-spatial bulk or single cell transcriptomics analysis approaches such as differential gene expression (DGE) analysis, comparing the mean expression of genes between two datasets by fold-change. However, this type of DGE analysis does not consider the spatial information. As such, a gene can be identified as not differentially expressed across two datasets by having the same mean gene expression, but the gene can still have two distinct spatial patterns of expression. On the other hand, a gene can be identified as differentially expressed across the two datasets by having very different mean gene expression but have spatial expression patterns that are highly similar. By not considering spatial information, traditional DGE analysis fails to characterize such cases. To demonstrate these cases, we simulated ST datasets where one pair of genes has different radial expression patterns but the same mean expression, while the other pair has same radial expression patterns but different mean expression. 

STcompare enables two differential spatial comparison tests, one based on spatial correlation, and one based on spatial fold change. 

# Installation 

(should be in the same style as SEraster)

# Tutorial 


## Load Libraries 

```{r setup, message=FALSE, warning=FALSE}
library(STcompare)
library(SpatialExperiment)
library(SEraster)
library(ggplot2)

```

## Load Dataset 
```{r load_data}

data("speKidney") 
```

## Simulated Pattern Kidney Example 

`speKidney` is a dataset of 3 simulated kidneys, where kidney A and B are simulated to have the same expression magnitude but different expression patterns and kidney A and C have the same expression patterns but different expression magnitudes.  

First, we use `SEraster` to rasterize this list of dataframes onto the same coordinate plane. 

```{r rasterization,  message=FALSE, warning=FALSE}
rastKidney <- SEraster::rasterizeGeneExpression(speKidney,
                assay_name = 'counts', resolution = 0.2,
                square = FALSE)

pA <- plotRaster(rastKidney$A, plotTitle = "Simulated Kidney A")
pA

pB <- plotRaster(rastKidney$B, plotTitle = "Simulated Kidney B")
pB

pC <- plotRaster(rastKidney$C, plotTitle = "Simulated Kidney C")
pC


```

### Mean comparision is not sufficient 

Traditionally, these simulated kidneys would be compared using the average of the mean expression. 

```{r box_plot,  message=FALSE, warning=FALSE}

df <- data.frame(
  value = c(
    as.numeric(assay(rastKidney$A, "pixelval")), 
    as.numeric(assay(rastKidney$B, "pixelval")), 
    as.numeric(assay(rastKidney$C, "pixelval"))
  ),
  sample = factor(rep(c("A", "B", "C"),
                      times = c(
                        length(as.numeric(assay(rastKidney$A, "pixelval"))), 
                        length(as.numeric(assay(rastKidney$B, "pixelval"))), 
                        length(as.numeric(assay(rastKidney$C, "pixelval")))
                      )
                  ))
)

compare_box_plot <- ggplot(df, aes(x = sample, y = value, fill = sample)) +
  geom_boxplot(outlier.shape = 16, outlier.size = 1.5, alpha = 0.7) +
  labs(x = "Sample", y = "Pixel Intensity", title = "Expression Magnitude Across Samples") +
  theme_minimal(base_size = 14) +
  theme(legend.position = "none")
compare_box_plot


```

Comparing the means allows us to differentiate between A and C since the expression magnitudes have a fold difference. 

However, comparing the mean does not capture the spatial gene expression difference between A and B. 

Using mean comparison is not sufficient to differentiate between gene expression patterns. 

## Using STcompare 

### Spatial fold change 

`spatialSimilarity` is used to compare fold change. 

```{r spatial_similarity,  message=FALSE, warning=FALSE}

# compare kidney A and B 
sAB <- spatialSimilarity(
  list(rastKidney$A, rastKidney$B)
)
  
# compare kidney A and C 
sAC <- spatialSimilarity(
  list(rastKidney$A, rastKidney$C)
)
  
# compare kidney B and C 
sBC <- spatialSimilarity(
  list(rastKidney$B, rastKidney$C)
)

```


53.5% of A and B pixels are within 2-fold difference of each other, while the rest are not. 

We see here that the inner radius is more highly expressed in B, while the outer radius is more highly expressed in C. 
The intermediate radius is shared between the two. 

```{r lr_pc_AB,  message=FALSE, warning=FALSE}

# linear regression and pixel classification plots 

lrAB <- linearRegression(input=sAB, gene = "Gene")
lrAB

pcAB <- pixelClass(input=sAB, gene="Gene")
pcAB

```


All the pixels in A are expressed over 2-folds in C. 

```{r lr_pc_AC,  message=FALSE, warning=FALSE}

lrAC <- linearRegression(input=sAC, gene = "Gene")
lrAC

pcAC <- pixelClass(input=sAC, gene="Gene")
pcAC
```


25.4% of B and C pixels are within 2-fold difference of each other, while the rest are not. 
We see here that the inner radius is similar expression magnitudes, while the outer radius is more highly expressed in C. 

```{r lr_pc_BC,  message=FALSE, warning=FALSE}

lrBC <- linearRegression(input=sBC, gene = "Gene")
lrBC

pcBC <- pixelClass(input=sBC, gene="Gene")
pcBC


```


## Spatial correlation 

We have used `spatialSimilarity` to understand the magnitude changes cross paired pixels. 

Now, we will use `spatialCorrelation` to understand correlation of expression across pixels. 


Gene expression at matched pixel locations are negatively correlated in A and B. 

```{r spatialCorrelationAB,  message=FALSE, warning=FALSE}

pixel_intersect_AB <- intersect(names(assays(rastKidney$A)$pixelval[1, ]), 
                                names(assays(rastKidney$B)$pixelval[1, ]))

scAB <- spatialCorrelation(
  X = assays(rastKidney$A)$pixelval[1, pixel_intersect_AB], 
  Y = assays(rastKidney$B)$pixelval[1, pixel_intersect_AB], 
  pos = spatialCoords(rastKidney$A)[pixel_intersect_AB, ]
)


scAB

rastGexpListAB <- list(A = rastKidney$A, B = rastKidney$B)
expAB <- plotCorrelationGeneExp(rastGexpListAB, scAB, "Gene")
expAB

```

Gene expression at matched pixel locations are positively correlated in A and C. 


```{r spatialCorrelationAC,  message=FALSE, warning=FALSE}

pixel_intersect_AC <- intersect(names(assays(rastKidney$A)$pixelval[1, ]), 
                                names(assays(rastKidney$C)$pixelval[1, ]))

scAC <- spatialCorrelation(
  X = assays(rastKidney$A)$pixelval[1, pixel_intersect_AC], 
  Y = assays(rastKidney$C)$pixelval[1, pixel_intersect_AC], 
  pos = spatialCoords(rastKidney$A)[pixel_intersect_AC, ]
)


scAC

rastGexpListAC <- list(A = rastKidney$A, C = rastKidney$C)
expAC <- plotCorrelationGeneExp(rastGexpListAC, scAC, "Gene")
expAC

```


Gene expression at matched pixel locations are negatively correlated in B and C. 

```{r spatialCorrelationBC,  message=FALSE, warning=FALSE}

pixel_intersect_BC <- intersect(names(assays(rastKidney$B)$pixelval[1, ]), 
                                names(assays(rastKidney$C)$pixelval[1, ]))

scBC <- spatialCorrelation(
  X = assays(rastKidney$B)$pixelval[1, pixel_intersect_BC], 
  Y = assays(rastKidney$C)$pixelval[1, pixel_intersect_BC], 
  pos = spatialCoords(rastKidney$B)[pixel_intersect_BC, ]
)


scBC

rastGexpListBC <- list(B = rastKidney$B, C = rastKidney$C)
expBC <- plotCorrelationGeneExp(rastGexpListBC, scBC, "Gene")
expBC


```


### Why does p-value correction matter 

`simRanPatternRasts` is a list of 100 simulated \code{SpatialExperiment} objects representing kidney-shaped 
datasets, each containing one independently generated spatially patterned gene with
no correlation between datasets. Each dataset consists of \eqn{N = 5000} simulated
cells distributed within a kidney-shaped region, with spatial coordinates and
expression values generated from Gaussian random fields.

Each dataset in `simRanPatternRasts` is simulated to have no correlation with each other. 

```{r eval=FALSE, echo=TRUE}

# This code was run once to generate the precomputed vignette results.
# Full script available at:
# system.file("scripts", "run_simRanPattern_analysis.R", package = "STcompare")

set.seed(0)
t <- Sys.time()

pvalueCorrected <- do.call(rbind, lapply(1:length(simRanPatternRasts), function(i) {
  sapply(1:length(simRanPatternRasts), function(j) {

    # shared pixels between rasters i and j
    vi <- intersect(rownames(spatialCoords(simRanPatternRasts[[i]])),
                    rownames(spatialCoords(simRanPatternRasts[[j]])))

    g1 <- SummarizedExperiment::assay(simRanPatternRasts[[i]], "pixelval")[1, vi]
    g2 <- SummarizedExperiment::assay(simRanPatternRasts[[j]], "pixelval")[1, vi]

    test <- cor.test(g1, g2)

    if (i != j) {
      rastGexpListAB <- list(i = simRanPatternRasts[[i]],
                             j = simRanPatternRasts[[j]])
      results <- spatialCorrelationGeneExp(
        rastGexpListAB,
        nPermutations = 100,
        nThreads = 5,
        BPPARAM = BiocParallel::MulticoreParam()
      )
      return(results$pValuePermuteX)
    } else {
      return(test$p.value)
    }
  })
}))
diag(pvalueCorrected) <- NA
tf <- Sys.time() - t
print(tf)

cors <- matrix(NA, nrow = length(simRanPatternRasts), ncol = length(simRanPatternRasts))
corspv <- cors

for (i in 1:length(simRanPatternRasts)) {
  for (j in 1:length(simRanPatternRasts)) {
    vi <- intersect(rownames(spatialCoords(simRanPatternRasts[[i]])),
                    rownames(spatialCoords(simRanPatternRasts[[j]])))

    g1 <- SummarizedExperiment::assay(simRanPatternRasts[[i]], "pixelval")[1, vi]
    g2 <- SummarizedExperiment::assay(simRanPatternRasts[[j]], "pixelval")[1, vi]

    ctest <- cor.test(g1, g2)
    cors[i, j] <- ctest$estimate
    corspv[i, j] <- ctest$p.value
  }
}

library(reshape2)
cors.df <- reshape2::melt(cors, value.name = "cors")
corspv.df <- reshape2::melt(corspv, value.name = "corspv")
pvalueCorrected.df <- reshape2::melt(pvalueCorrected, value.name = "corspv_corrected")

cors_df <- data.frame(cors.df,
                      corspv = corspv.df$corspv,
                      corspv_corrected = pvalueCorrected.df$corspv_corrected)



cors_df$corspv_corrected[cors_df$corspv_corrected == 0] <- 0.01



corrected <- ggplot(cors_df) +
  geom_point(aes(x = cors, y = -log10(corspv)),
             alpha = 0.1, size = 0.5, color = "blue") +
  geom_point(aes(x = cors, y = -log10(corspv_corrected)),
             alpha = 0.1, size = 0.5, color = "green") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  theme_classic() +
  labs(x = "Correlation", y = "-log10(p-value)",
       title = "Naïve vs Spatially Corrected p-values") +
  ylim(min(-log10(cors_df$corspv), na.rm = TRUE),
       max(-log10(cors_df$corspv), na.rm = TRUE))
corrected


```




```{r load-results, eval=FALSE, echo=TRUE, message=FALSE}

data_path <- system.file("extdata", "simRanPatternResults.RData", package = "STcompare")
load(data_path)
corrected

```

Without p-value correction, we observe that 50% of these p-values are less than 0.05, 
while after correcting for p-value only 4% of the empirical p-values are less than 0.05. 

# Conclusion 

Changes in expression magnitude and pattern can be independent observations. 

- A and B have similar magnitude (no fold change) and negatively correlated 
- A and C not similar magnitude and are positively correlated
- B and C not similar magnitude and are negatively correlated

These are differences that would otherwise have been able to be detected using mean gene expression magnitude comparison. 

P value correction is needed to take into account that gene expression in each pixel is not independent from that of its neighbors’. 











