<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started with STcompare • STcompare</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with STcompare">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">STcompare</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting-started-with-STcompare_old.html">Getting Started with STcompare</a></li>
    <li><a class="dropdown-item" href="../articles/getting-started-with-STcompare.html">Getting Started with STcompare</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting Started with STcompare</h1>
                        <h4 data-toc-skip class="author">Vivien
Jiang</h4>
                        <h4 data-toc-skip class="author">Kalen
Clifton</h4>
                        <h4 data-toc-skip class="author">Jean Fan</h4>
            
      

      <div class="d-none name"><code>getting-started-with-STcompare_old.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introducton">Introducton<a class="anchor" aria-label="anchor" href="#introducton"></a>
</h2>
<p>Spatial transcriptomic (ST) technologies enable the investigation of
how tissue organization relates to cellular function by profiling the
spatial location of cells within a tissue and the gene expression
profiles associated with those locations. As ST technologies are
increasingly applied in the context of disease characterization and
translational research, identifying genes that spatially change in their
expression patterns in diseased tissues as compared to healthy controls
could reveal localized molecular variation associated with pathological
processes, offering insights into disease mechanisms as well as aiding
in the discovery of diagnostic biomarkers.</p>
<p>Attempts at such comparative analysis of ST datasets can be performed
using traditional non-spatial bulk or single cell transcriptomics
analysis approaches such as differential gene expression (DGE) analysis,
comparing the mean expression of genes between two datasets by
fold-change. However, this type of DGE analysis does not consider the
spatial information. As such, a gene can be identified as not
differentially expressed across two datasets by having the same mean
gene expression, but the gene can still have two distinct spatial
patterns of expression. On the other hand, a gene can be identified as
differentially expressed across the two datasets by having very
different mean gene expression but have spatial expression patterns that
are highly similar. By not considering spatial information, traditional
DGE analysis fails to characterize such cases. To demonstrate these
cases, we simulated ST datasets where one pair of genes has different
radial expression patterns but the same mean expression, while the other
pair has same radial expression patterns but different mean
expression.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'JEFworks-Lab/STcompare'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorial">Tutorial<a class="anchor" aria-label="anchor" href="#tutorial"></a>
</h2>
<p>In this tutorial we will walk through applying
<code>STcompare</code>’s two differential spatial comparison tests to
simulated kidney datasets:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Spatial Correlation:</strong> Person correlation assumes
that each sample is independent and identically distributed random
variables. However, in terms of gene expression, the gene expression of
one cell influences the gene expression of the neighboring cells.
Spatial Correlation computes the empirical p-value instead to consider
the dependent relationship. We will show why the p-value null
distribution correction is needed in the Spatial Correlation
test.</p></li>
<li><p><strong>Spatial Fold Change:</strong> Computes a Similarity score
for each gene in the comparison, compares the change in expression at
matched locations.</p></li>
</ol>
<div class="section level3">
<h3 id="load-libraries">Load Libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jef.works/STcompare/">STcompare</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SEraster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="simulated-pattern-kidney">Simulated Pattern Kidney<a class="anchor" aria-label="anchor" href="#simulated-pattern-kidney"></a>
</h3>
<p>Here we load in a simulated kidney dataset that we will use as an
example dataset for the rest of the tutorial.</p>
<p><a href="%22https://www.bioconductor.org/packages/release/bioc/html/SpatialExperiment.html%22" class="external-link">‘SpatialExperiment’</a>
objects store spatial transcriptomics data, combining gene expression
matrices with the spatial coordinates of each spot or cell.
<code>speKidney</code> is a list of three <code>SpatialExperiment</code>
objects representing simulated kidneys, where A and B share expression
magnitude but differ in spatial pattern, and A and C share spatial
pattern but differ in magnitude.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"speKidney"</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">speKidney</span><span class="op">)</span></span>
<span><span class="co">#&gt; $A</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 1 1229 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(1): Gene</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1229): cell1 cell2 ... cell1228 cell1229</span></span>
<span><span class="co">#&gt; colData names(1): sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $C</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 1 1297 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(1): Gene</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1297): cell1 cell2 ... cell1296 cell1297</span></span>
<span><span class="co">#&gt; colData names(1): sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $B</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 1 1242 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(1): Gene</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1242): cell1 cell2 ... cell1241 cell1242</span></span>
<span><span class="co">#&gt; colData names(1): sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span></span>
<span><span class="va">speA</span> <span class="op">&lt;-</span> <span class="va">speKidney</span><span class="op">$</span><span class="va">A</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">speA</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">speA</span><span class="op">)</span><span class="op">)</span>        </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">speA</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="st">"Gene"</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">Gene</span> <span class="op">&lt;-</span> <span class="va">gene_expr</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">Gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"x"</span>, y <span class="op">=</span> <span class="st">"y"</span>, title <span class="op">=</span> <span class="st">"Kidney A"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/load_data-1.png" width="576"></p>
<p>Each element of <code>speKidney</code> contains a gene-expression
assay, x–y spatial coordinates for each spot, and associated metadata.
The three samples vary slightly in the number of spatial locations but
share the same structure.</p>
<p>First, we use <code>SEraster</code> to rasterize this list of
simulated kidneys, stored as <code>SpatialExperiment</code> objects,
onto the same coordinate plane.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rastKidney</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SEraster/man/rasterizeGeneExpression.html" class="external-link">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">speKidney</span>,</span>
<span>                assay_name <span class="op">=</span> <span class="st">'counts'</span>, resolution <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                square <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># After rasterization, the output is a SpatialExperiment object, but the spatial units are now raster pixels rather than individual cells or spots. The assay has been converted to pixelval, and additional metadata (num_cell, cellID_list, geometry) records which original cells contributed to each pixel.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rastKidney</span><span class="op">)</span></span>
<span><span class="co">#&gt; $A</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 1 282 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): pixelval</span></span>
<span><span class="co">#&gt; rownames(1): Gene</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(282): pixel52 pixel53 ... pixel416 pixel417</span></span>
<span><span class="co">#&gt; colData names(6): num_cell cellID_list ... geometry sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $C</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 1 287 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): pixelval</span></span>
<span><span class="co">#&gt; rownames(1): Gene</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(287): pixel51 pixel52 ... pixel417 pixel431</span></span>
<span><span class="co">#&gt; colData names(6): num_cell cellID_list ... geometry sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $B</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 1 279 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): pixelval</span></span>
<span><span class="co">#&gt; rownames(1): Gene</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(279): pixel53 pixel65 ... pixel416 pixel417</span></span>
<span><span class="co">#&gt; colData names(6): num_cell cellID_list ... geometry sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span></span>
<span><span class="co"># These are the plots to visualize what the kidneys looks like </span></span>
<span><span class="va">pA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SEraster/man/plotRaster.html" class="external-link">plotRaster</a></span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">A</span>, plotTitle <span class="op">=</span> <span class="st">"Simulated Kidney A"</span><span class="op">)</span></span>
<span><span class="va">pA</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/rasterization-1.png" width="576"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SEraster/man/plotRaster.html" class="external-link">plotRaster</a></span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">B</span>, plotTitle <span class="op">=</span> <span class="st">"Simulated Kidney B"</span><span class="op">)</span></span>
<span><span class="va">pB</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/rasterization-2.png" width="576"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SEraster/man/plotRaster.html" class="external-link">plotRaster</a></span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">C</span>, plotTitle <span class="op">=</span> <span class="st">"Simulated Kidney C"</span><span class="op">)</span></span>
<span><span class="va">pC</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/rasterization-3.png" width="576"></p>
<div class="section level4">
<h4 id="mean-comparision-is-not-sufficient-d">Mean comparision is not sufficient d<a class="anchor" aria-label="anchor" href="#mean-comparision-is-not-sufficient-d"></a>
</h4>
<p>Traditionally, these simulated kidneys would be compared using the
average of the mean expression.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Convert the rasterized pixel values for each kidney (A, B, C) into one dataframe.</span></span>
<span><span class="co"># Each kidney’s pixel values are extracted from the "pixelval" assay and stored as a numeric vector.</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">A</span>, <span class="st">"pixelval"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">B</span>, <span class="st">"pixelval"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">C</span>, <span class="st">"pixelval"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co"># Create a "sample" column showing which kidney each pixel belongs to.</span></span>
<span>  <span class="co"># The `rep()` call ensures that the correct number of labels (A/B/C) is assigned</span></span>
<span>  <span class="co"># based on the length of each pixel value vector.</span></span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">A</span>, <span class="st">"pixelval"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">B</span>, <span class="st">"pixelval"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">rastKidney</span><span class="op">$</span><span class="va">C</span>, <span class="st">"pixelval"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot a boxplot comparing the distribution of pixel intensities across the 3 kidneys.</span></span>
<span><span class="co"># This represents a traditional magnitude-only comparison, </span></span>
<span><span class="co"># which cannot distinguish pattern similarity/differences.</span></span>
<span><span class="va">compare_box_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.shape <span class="op">=</span> <span class="fl">16</span>, outlier.size <span class="op">=</span> <span class="fl">1.5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Sample"</span>, y <span class="op">=</span> <span class="st">"Pixel Intensity"</span>, title <span class="op">=</span> <span class="st">"Expression Magnitude Across Samples"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">compare_box_plot</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/box_plot-1.png" width="576"></p>
<p>Comparing the means allows us to differentiate between A and C since
the expression magnitudes have a fold difference. However, comparing the
mean does not capture the spatial gene expression difference between A
and B.</p>
<ol style="list-style-type: decimal">
<li>We will use <code>spatialCorrelationGeneExp</code> test to quantify
the correlation in gene expression pattern between (A and B) and (A and
C). Where (A and B) will have a significant negative correlation and (A
and C) will have a significant postive correation.</li>
<li>And we will use <code>SpatialSimilarity</code> test quantify the
difference in gene expression magnitude not reflected by DGE between (A
and B)</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="using-stcompare">Using STcompare<a class="anchor" aria-label="anchor" href="#using-stcompare"></a>
</h3>
<div class="section level4">
<h4 id="input-formatting">Input formatting<a class="anchor" aria-label="anchor" href="#input-formatting"></a>
</h4>
<ol style="list-style-type: decimal">
<li><p><strong>Tissue Alignment</strong><br>
For <code>STcompare</code> to produce meaningful comparisons, the
tissues must first be spatially aligned so that corresponding structures
are matched across samples. The <code>STalign</code> package can be used
to align two tissues. <a href="%22https://jef.works/STalign/notebooks/merfish-merfish-alignment.html%22" class="external-link">STalign
Tutorial</a></p></li>
<li><p><strong>Rasterization</strong><br><code>STcompare</code> takes a list of <a href="%22https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html%22" class="external-link"><code>SpatialExperiment</code></a>
objects and requires them to be on a shared coordinate plane. Use
<code>SEraster</code> to rasterize multiple samples onto the same plane,
allowing <code>STcompare</code> to pairwise compare across any number of
samples. <a href="%22https://jef.works/SEraster/articles/formatting-SpatialExperiment-for-SEraster.html%22" class="external-link">SEraster
formatting tutorial</a></p></li>
</ol>
</div>
<div class="section level4">
<h4 id="spatial-correlation">Spatial correlation<a class="anchor" aria-label="anchor" href="#spatial-correlation"></a>
</h4>
<p>Now, we will use <code>spatialCorrelationGeneExp</code> to understand
correlation of expression across pixels.</p>
<ul>
<li>
<strong>correlationCoef</strong> Pearson’s is correlation
coefficient shows the strength and direction of a linear relationship
between the two objects</li>
<li>
<strong>pValueNaive</strong> is the analytical p-value naively
assuming independent observations, often time not accurate<br>
</li>
<li>
<strong>pValuePermuteY</strong> is the p-value when creating an
empirical null from permutations of observations in Y – the higher of
<strong>pValuePermuteY</strong>, <strong>pValuePermuteX</strong>, is
more accurate p-value due to corrected null distribution</li>
<li>
<strong>pValuePermuteX</strong> the p-value when creating an
empirical null from permutations of observations in X</li>
<li>
<strong>deltaStarMedianX</strong> the median delta star (the delta
which minimizes the difference between the variogram of the permutation
and the variogram of observations) across permutations of X</li>
<li>
<strong>deltaStarMedianY</strong> the median delta star across
permutations of Y</li>
<li>
<strong>deltaStarXlist</strong> of delta star for all permutations
of X</li>
<li>
<strong>deltaStarY</strong> is list of delta star for all
permutations of Y</li>
<li>
<strong>nullCorrelationsX</strong> correlation coefficients for Y
and all permuations of X</li>
<li>
<strong>nullCorrelationsY</strong> correlation coefficients for X
and all permuations of Y</li>
</ul>
<p>Gene expression at matched pixel locations are negatively correlated
in A and B.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># From the list of rasterized kidneys, rastKidney, we will take a subset of rastKidney </span></span>
<span><span class="co"># rastGexpListAB is a list of two kidneys, A and B </span></span>
<span><span class="va">rastGexpListAB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>A <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">A</span>, B <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">B</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># spatialCorrelationGeneExp takes input of a list of two SpatialExperiment objects -- rastGexpListAB</span></span>
<span><span class="co"># nThreads, the default is 1, should be set to the number of cores available to allow for parallel computing </span></span>
<span><span class="va">scAB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialCorrelationGeneExp.html">spatialCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpListAB</span>, nThreads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># correlationCoef is showing a negative linear relationship </span></span>
<span><span class="co"># Both the naive and the permuted p-values (pValuePermuteY and pValuePermuteX) are showing that the correlation is significant </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scAB</span><span class="op">)</span></span>
<span><span class="co">#&gt;      correlationCoef   pValueNaive pValuePermuteX pValuePermuteY</span></span>
<span><span class="co">#&gt; Gene      -0.9472813 5.652003e-136              0              0</span></span>
<span><span class="co">#&gt;      deltaStarMedianX deltaStarMedianY   deltaStarX   deltaStarY</span></span>
<span><span class="co">#&gt; Gene              0.2              0.2 0.1, 0.1.... 0.5, 0.2....</span></span>
<span><span class="co">#&gt;      nullCorrelationsX nullCorrelationsY</span></span>
<span><span class="co">#&gt; Gene      0.119003....      0.087039....</span></span>
<span></span>
<span><span class="co"># visualization of the negative correlation </span></span>
<span><span class="co"># plotCorrelationGeneExp needs the list of rasterized SpatialExperiment objects, the table from spatialCorrelationGeneExp of the same objects, </span></span>
<span><span class="co"># and the gene name you are trying to plot. In the case, the gene name is "Gene". </span></span>
<span><span class="va">expAB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCorrelationGeneExp.html">plotCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpListAB</span>, <span class="va">scAB</span>, <span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">expAB</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/spatialCorrelationA-1.png" width="576"></p>
<p>Gene expression at matched pixel locations are positively correlated
in A and C.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># From the list of rasterized kidneys, rastKidney, we will take a subset of rastKidney </span></span>
<span><span class="co"># rastGexpListAC is a list of two kidneys, A and C</span></span>
<span><span class="va">rastGexpListAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>A <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">A</span>, C <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">C</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># correlationCoef is showing a positive linear relationship </span></span>
<span><span class="co"># Both the naive and the permuted p-values (pValuePermuteY and pValuePermuteX) are showing that the correlation is significant </span></span>
<span><span class="va">scAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialCorrelationGeneExp.html">spatialCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpListAC</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scAC</span><span class="op">)</span></span>
<span><span class="co">#&gt;      correlationCoef   pValueNaive pValuePermuteX pValuePermuteY</span></span>
<span><span class="co">#&gt; Gene       0.9431531 1.409195e-133              0              0</span></span>
<span><span class="co">#&gt;      deltaStarMedianX deltaStarMedianY   deltaStarX   deltaStarY</span></span>
<span><span class="co">#&gt; Gene              0.2              0.2 0.4, 0.3.... 0.4, 0.2....</span></span>
<span><span class="co">#&gt;      nullCorrelationsX nullCorrelationsY</span></span>
<span><span class="co">#&gt; Gene      0.201840....      0.036416....</span></span>
<span></span>
<span><span class="co"># visualization of the positive correlation</span></span>
<span><span class="co"># plotCorrelationGeneExp needs the list of rasterized SpatialExperiment objects, the table from spatialCorrelationGeneExp of the same objects, </span></span>
<span><span class="co"># and the gene name you are trying to plot. In the case, the gene name is "Gene".</span></span>
<span><span class="va">expAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCorrelationGeneExp.html">plotCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpListAC</span>, <span class="va">scAC</span>, <span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">expAC</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/spatialCorrelationAC-1.png" width="576"></p>
<p>Gene expression at matched pixel locations are negatively correlated
in B and C.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="co"># From te list of rasterized kidneys, rastKidney, we will take a subset of rastKidney </span></span>
<span><span class="co"># rastGexpListAC is a list of two kidneys, B and C</span></span>
<span><span class="va">rastGexpListBC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>B <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">B</span>, C <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">C</span><span class="op">)</span></span>
<span><span class="va">scBC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialCorrelationGeneExp.html">spatialCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpListBC</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scBC</span><span class="op">)</span></span>
<span><span class="co">#&gt;      correlationCoef   pValueNaive pValuePermuteX pValuePermuteY</span></span>
<span><span class="co">#&gt; Gene      -0.9478025 4.914519e-138              0              0</span></span>
<span><span class="co">#&gt;      deltaStarMedianX deltaStarMedianY   deltaStarX   deltaStarY</span></span>
<span><span class="co">#&gt; Gene              0.2              0.2 0.2, 0.3.... 0.2, 0.3....</span></span>
<span><span class="co">#&gt;      nullCorrelationsX nullCorrelationsY</span></span>
<span><span class="co">#&gt; Gene      -0.34439....      0.221883....</span></span>
<span></span>
<span><span class="co"># visualization of the negative correlation </span></span>
<span><span class="co"># plotCorrelationGeneExp needs the list of rasterized SpatialExperiment objects, the table from spatialCorrelationGeneExp of the same objects, </span></span>
<span><span class="co"># and the gene name you are trying to plot. In the case, the gene name is "Gene". </span></span>
<span><span class="va">expBC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCorrelationGeneExp.html">plotCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpListBC</span>, <span class="va">scBC</span>, <span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">expBC</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/spatialCorrelationBC-1.png" width="576">
### Why does p-value correction matter</p>
<p><code>simRanPatternRasts</code> is a list of 100 simulated
<code>SpatialExperiment</code> objects representing kidney-shaped
datasets, each containing one independently generated spatially
patterned gene with no correlation between datasets. Each dataset
consists of N = 5000 simulated cells distributed within a kidney-shaped
region, with spatial coordinates and expression values generated from
Gaussian random fields. All it the objects in
<code>simRanPatternRasts</code> is already rasterized onto the same
coordinate plane.</p>
<p>Using the kidney 1 and kidney 7 in the
<code>simRanPatternRasts</code> dataset, we see that the p-value is in
the significant range (p-value &lt; 0.05). However, based on the way the
kidneys are simulated (simulated to have no correlation), this should
not be the case, which is reflected in the permuted p-values. The unlike
the naive p-value, the permuted p-values do not show significant
correlation.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># using plotRaster from the SEraster package to visualize kidney 1 and 7</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SEraster/man/plotRaster.html" class="external-link">plotRaster</a></span><span class="op">(</span><span class="va">simRanPatternRasts</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, plotTitle <span class="op">=</span> <span class="st">"simRanPatternRasts Kidney 1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/show_ran_kid-1.png" width="576"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SEraster/man/plotRaster.html" class="external-link">plotRaster</a></span><span class="op">(</span><span class="va">simRanPatternRasts</span><span class="op">[[</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span>, plotTitle <span class="op">=</span> <span class="st">"simRanPatternRasts Kidney 7"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/show_ran_kid-2.png" width="576"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># taking a subset of simRanPatternRasts</span></span>
<span><span class="co"># rastGexpList is a list of rasterize SpatialExperiment objects kidney 1 and kidney 7 </span></span>
<span><span class="va">rastGexpList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>kidney_1 <span class="op">=</span> <span class="va">simRanPatternRasts</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, kidney_7 <span class="op">=</span> <span class="va">simRanPatternRasts</span><span class="op">[[</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># finding the correlation and the p-value of kidney 1 and kidney 7 </span></span>
<span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialCorrelationGeneExp.html">spatialCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpList</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the naive p-value is showing that the correlation is significant </span></span>
<span><span class="co"># while both of the permuted p-values are showing not significant p-values </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span> </span>
<span><span class="co">#&gt;   correlationCoef  pValueNaive pValuePermuteX pValuePermuteY deltaStarMedianX</span></span>
<span><span class="co">#&gt; 1      -0.2189486 0.0002533931           0.29           0.19              0.1</span></span>
<span><span class="co">#&gt;   deltaStarMedianY   deltaStarX   deltaStarY nullCorrelationsX</span></span>
<span><span class="co">#&gt; 1             0.25 0.2, 0.1.... 0.9, 0.1....      0.044322....</span></span>
<span><span class="co">#&gt;   nullCorrelationsY</span></span>
<span><span class="co">#&gt; 1      0.044163....</span></span>
<span></span>
<span><span class="co"># the plot further shows that there isn't a correlation between kidney 1 and kidney 7</span></span>
<span><span class="fu"><a href="../reference/plotCorrelationGeneExp.html">plotCorrelationGeneExp</a></span><span class="op">(</span><span class="va">rastGexpList</span>, <span class="va">sc</span>, <span class="st">"1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/show_ran_kid-3.png" width="576"></p>
<p>For each pairwise kidney SpatialExperiment object in the list, the
corrected p-value is computed using
<code>spatialCorrelationGeneExp</code></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># This code was run once to generate the precomputed vignette results.</span></span>
<span><span class="co"># Full script available at:</span></span>
<span><span class="co"># system.file("scripts", "simRanPatternSpatialCorrelation.R", package = "STcompare")</span></span>
<span></span>
<span><span class="va">data_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"simRanPatternResults.RData"</span>, package <span class="op">=</span> <span class="st">"STcompare"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">data_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Var1 and Var2 is every pairwise combination between the 100 kidneys </span></span>
<span><span class="co"># cors is the correlation of that pair </span></span>
<span><span class="co"># corspv is the naive p-value for that pair </span></span>
<span><span class="co"># corspv_correct is the permuted p-value for that pair (chosen to be the higher of pValuePermuteY and pValuePermuteX)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cors_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   Var1 Var2        cors     corspv corspv_corrected</span></span>
<span><span class="co">#&gt; 1    1    1  1.00000000 0.00000000               NA</span></span>
<span><span class="co">#&gt; 2    2    1 -0.04158115 0.49227077             0.85</span></span>
<span><span class="co">#&gt; 3    3    1  0.09066560 0.13439158             0.56</span></span>
<span><span class="co">#&gt; 4    4    1  0.05189013 0.39310096             0.80</span></span>
<span><span class="co">#&gt; 5    5    1  0.11116927 0.06515094             0.45</span></span>
<span><span class="co">#&gt; 6    6    1  0.10915193 0.07229390             0.47</span></span>
<span></span>
<span><span class="co"># remove the na values </span></span>
<span><span class="va">cors_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">cors_df</span><span class="op">)</span></span>
<span></span>
<span><span class="va">corrected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cors_df</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span><span class="va">cors</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">corspv</span><span class="op">)</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.1</span>, size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span><span class="va">cors</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">corspv_corrected</span><span class="op">)</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.1</span>, size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Correlation"</span>, y <span class="op">=</span> <span class="st">"-log10(p-value)"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"Naïve vs Spatially Corrected p-values"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">'dashed'</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">cors_df</span><span class="op">$</span><span class="va">corspv</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">cors_df</span><span class="op">$</span><span class="va">corspv</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Correlation"</span> , y <span class="op">=</span> <span class="st">"-log10(p-value)"</span><span class="op">)</span></span>
<span><span class="va">corrected</span></span>
<span><span class="co">#&gt; Warning: Removed 85 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_point()`).</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/load-results-1.png" width="576"></p>
<p>Without p-value correction, we observe that 50% of these p-values are
less than 0.05, while after correcting for p-value only 4% of the
empirical p-values are less than 0.05.</p>
<p>Therefore, the permuted p-value is needed to more accurately capture
true correlation significance.</p>
</div>
<div class="section level4">
<h4 id="spatial-fold-change">Spatial fold change<a class="anchor" aria-label="anchor" href="#spatial-fold-change"></a>
</h4>
<p>Now, we have use <code>spatialSimilarity</code> to understand the
magnitude changes cross paired pixels.</p>
<p>Here, we find the spatialSimilarity pairwise between in the
<code>speKidney</code> between (A, B, C)</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># spatialSimilarity also takes in a list of 2 SpatialExperiment objects you want to compare between </span></span>
<span><span class="co"># t1 and t2 are the thresholding value for the first and second object in the list. </span></span>
<span><span class="co"># The threshold is used to remove pixels that have no or little expression for a given gene in the pixel.</span></span>
<span><span class="co"># Failure to remove those those pixels will result in a trivial similarity score due to noise. </span></span>
<span><span class="co"># If t1 and t2 aren't provided, minQuantile is used to threshold based on the quantile of expression. </span></span>
<span><span class="co"># The default for minQuantile is 0.05, meaning by default the pixels at the bottom 5% are removed. </span></span>
<span><span class="co"># minPixels is the percentage of pixels left after thresholding. The default is 0.1. </span></span>
<span><span class="co"># When there is not enough pixels left, the similarity score will also be trivial.</span></span>
<span><span class="co"># foldChange is the number of fold that are considered similar. The default is 1 fold. </span></span>
<span></span>
<span><span class="co"># compare kidney A and B </span></span>
<span><span class="va">sAB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialSimilarity.html">spatialSimilarity</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>A <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">A</span>, B <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">B</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compare kidney A and C </span></span>
<span><span class="va">sAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialSimilarity.html">spatialSimilarity</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>A <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">A</span>, C <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">C</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># compare kidney B and C </span></span>
<span><span class="va">sBC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialSimilarity.html">spatialSimilarity</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>B <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">B</span>, C <span class="op">=</span> <span class="va">rastKidney</span><span class="op">$</span><span class="va">C</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For gene <code>Gene</code> in the comparision between A and B, we see
that A and B has a similarity score of S=0.535. This means that 53.5% of
the intersecting pixels between A and B are within a 1</p>
<p>We see here that the inner radius is more highly expressed in B,
while the outer radius is more highly expressed in C. The intermediate
radius is shared between the two.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># To get the linear regression and the pixel classification plots, </span></span>
<span><span class="co"># the inputs are the spatialSimilarity output and the gene name </span></span>
<span></span>
<span><span class="co"># The Linear regression shows the areas in which gene expression </span></span>
<span><span class="co"># each pixels falls, the pixel is either: higher in B, similar, higher in A  </span></span>
<span><span class="va">lrAB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linearRegression.html">linearRegression</a></span><span class="op">(</span>input<span class="op">=</span><span class="va">sAB</span>, gene <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">lrAB</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/lr_pc_AB-1.png" width="576"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Similar to the linear regression plot, the pixel classification plot takes </span></span>
<span><span class="co"># the plot of the rasterized kidney and classifies each pixel with the pixels </span></span>
<span><span class="co"># that are below the threshold are gray </span></span>
<span><span class="va">pcAB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pixelClass.html">pixelClass</a></span><span class="op">(</span>input<span class="op">=</span><span class="va">sAB</span>, gene<span class="op">=</span><span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">pcAB</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/lr_pc_AB-2.png" width="576"></p>
<p>All the pixels have over 1-fold increase in C relative to the same
pixels in A.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># all the pixels have higher expression in C than in A, </span></span>
<span><span class="co"># and are outside of the similar expression boundaries</span></span>
<span></span>
<span><span class="va">lrAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linearRegression.html">linearRegression</a></span><span class="op">(</span>input<span class="op">=</span><span class="va">sAC</span>, gene <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">lrAC</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/lr_pc_AC-1.png" width="576"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pcAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pixelClass.html">pixelClass</a></span><span class="op">(</span>input<span class="op">=</span><span class="va">sAC</span>, gene<span class="op">=</span><span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">pcAC</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/lr_pc_AC-2.png" width="576"></p>
<p>25.4% of B and C pixels are within 1-fold difference of each other,
while the rest are not.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">lrBC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/linearRegression.html">linearRegression</a></span><span class="op">(</span>input<span class="op">=</span><span class="va">sBC</span>, gene <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">lrBC</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/lr_pc_BC-1.png" width="576"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># We see here that the inner radius is similar expression magnitudes, while the outer radius is more highly expressed in C. </span></span>
<span><span class="va">pcBC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pixelClass.html">pixelClass</a></span><span class="op">(</span>input<span class="op">=</span><span class="va">sBC</span>, gene<span class="op">=</span><span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">pcBC</span></span></code></pre></div>
<p><img src="getting-started-with-STcompare-figures/lr_pc_BC-2.png" width="576"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Changes in expression magnitude and pattern can be independent
observations.</p>
<ul>
<li>A and B have similar magnitude (no fold change) and negatively
correlated</li>
<li>A and C do not have similar magnitude and are positively
correlated</li>
<li>B and C do not have similar magnitude and are negatively
correlated</li>
</ul>
<p>These are differences that would otherwise have been able to be
detected using mean gene expression magnitude comparison. P value
correction is needed to take into account that gene expression in each
pixel is not independent from that of its neighbors’.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kalen Clifton, Vivien Jiang, Jean Fan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
